Prepare the Submission for Evaluation

mdsage1 opened the pull request:
This pull request addresses the need to prepare a config file for the submission after the compressed submission has been opened to allow the use of all files.

The steps are:
1. Unzip the compressed submission
2. Generate a config file using the paths to the de-identified images, uid and patid mapping CSV files
3. Run the steps in the following order:
![image](https://github.com/Sage-Bionetworks-Challenges/MIDI-B-De-identification/assets/122999770/a542ab23-5311-4e2b-b11b-e482fcdd7593)
**Please confirm whether I need to add the status check steps in between these to prevent them from running simultaneously**
4. Then validate the writeup (This may need to be changed as it currently checks to see if the writeup is a project which it won't be once the compressed submission is unzipped)

vpchung commented on steps/unzip_submission.py:
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
```
This script can be part of the `docker.synapse.org/syn53065762/validate_score:v4` docker model too, if you wanted.

mdsage1 replied:
Thanks I figured I put it in here so I had a reference in this repo about what I did to handle the need for a config file to be created for each submission.

vpchung commented on workflow.cwl:
```diff
@@ -92,20 +92,61 @@ steps:
-      - id: synapse_config
-        source: "#synapseConfig"
+      - id: compressed_file
+        source: "#download_compressed_file/filepath"
```
`download_compressed_file` does not currently exist as a step in the workflow.  Typo?  Or maybe a missed commit?

vpchung reviewed (changes requested):
This is a great start!  I think there may be some missing steps and scripts from the PR though, perhaps missed commits or pushes?

mdsage1 commented on workflow.cwl:
```diff
@@ -92,20 +92,61 @@ steps:
+      - id: status
+    
+  create_discrepancy_report:
+    run: steps/discrepancy.cwl
```
Great catch! I missed it when I pushed the changes but I've added it to the PR now!

mdsage1 commented on workflow.cwl:
```diff
@@ -92,20 +92,61 @@ steps:
+      - id: results
+      - id: status
+
+  create_dciovdfy_report:
```
Yes, the participants will recieve `dciovdfy_results`.

mdsage1 commented on workflow.cwl:
```diff
@@ -92,20 +92,61 @@ steps:
+      - id: config_json
+      - id: writeup_file
+
+  create_scoring_report:
```
create_scoring_report will generate a report for the participants. Can I just omit the other outputs though?

vpchung replied:
Yep!  You can remove any outputs (and inputs) that aren't needed or used.

mdsage1 commented on steps/unzip_submission.py:
```diff
@@ -0,0 +1,80 @@
+import json
+import zipfile
+
+def inspect_zip(zip_path):
```
I've added handling for tar files and a check.

mdsage1 commented on workflow.cwl:
```diff
@@ -92,20 +92,61 @@ steps:
+      - id: results
+      - id: status
+    
+  create_discrepancy_report:
```
Answered above

vpchung reviewed (changes requested):
Thanks for addressing the issues!  

A couple more things to think about:
* If the submission is not a zipfile or tarfile, how can we let the participant know? Because we are using TOIL to run a CWL workflow, the ValueError within the Python script will not be seen by the participant.  Instead, the logfile will say something along the lines of expected CWL output not found - which will probably not be helpful to the participant.  

   One possible way to address this is to do something [like this](https://github.com/Sage-Bionetworks-Challenges/olfactory-mixtures-prediction/blob/407e79f152bcd94b36813d7ead2fbf8110ba0967/workflow.cwl#L93-L139).  Here, I wrote [some inline Python code](https://github.com/Sage-Bionetworks-Challenges/olfactory-mixtures-prediction/blob/407e79f152bcd94b36813d7ead2fbf8110ba0967/steps/validate_docker.cwl) where it will check whether the participant submitted a Docker image.  If they did, then `status` will be set as VALID, otherwise, `status` will be INVALID.  If the `status` is INVALID, then the next workflow step (`notify_docker_status`) will email the participant to let them know why their submission failed.  Finally, the third step highlighted (`add_status_annots`) is to annotate the submission with the results, so that organizers can know what happened.
* Relatedly, if the submission is not a zipfile or tarfile, a step should be added so that the workflow doesn't continue to run.  Using the same workflow reference above, [here's the step](https://github.com/Sage-Bionetworks-Challenges/olfactory-mixtures-prediction/blob/407e79f152bcd94b36813d7ead2fbf8110ba0967/workflow.cwl#L141-L153) you can use for reference, where it uses the `status` output of the previous validation step to determine whether it should continue on to running the submission.
* You mentioned [here](https://github.com/Sage-Bionetworks-Challenges/MIDI-B-De-identification/pull/4#discussion_r1631523681) that participants will receive `dciovdfy_results` - are you thinking of uploading the file to Synapse so that they can have access to those results? Or will the results be put into a leaderboard?

   If uploading the file, then you will probably need to add another workflow step in order to upload it and return the synID as output.  That way, you can later email that synID to the participant.  Perhaps something like [this](https://github.com/Sage-Bionetworks-Challenges/brats2023/blob/main/evaluation/lesionwise/score.py#L116-L121), where a [reconfigured email step](https://github.com/Sage-Bionetworks-Challenges/brats2023/blob/main/shared/email_results.cwl) is used in order to notify the participant.

   If leaderboard, then the workflow will need an additional annotation step.  However, in order to annotate, the results will need to be in captured in a JSON file, since that is what [`annotate-submission`](https://sage-bionetworks.github.io/challengeutils/client/admin.html#annotate-a-submission) expects.  The easiest way to do this would probably to add another workflow step that will convert the CSV into JSON.

---

I know that was a lot of info all in one-go 😆🙃  If it's easier to manage, I can approve this PR now, and then you can focus on each of the above points in separate PRs.  Lmk!

mdsage1 commented on steps/discrepancy.cwl:
```diff
@@ -0,0 +1,40 @@
+    outputBinding:
+      glob: results.json
+
+  - id: status
```
I've commented out unused outputs that may be used later.

mdsage1 commented on steps/discrepancy.cwl:
```diff
@@ -0,0 +1,40 @@
+    outputBinding:
+      glob: discrepancy_report.csv
+
+  - id: results
```
I've commented out unused outputs that may be used later.

mdsage1 commented on steps/dicovdfy.cwl:
```diff
@@ -0,0 +1,40 @@
+    outputBinding:
+      glob: dciodvfy_report.csv
+
+  - id: results
```
I've commented out unused outputs that may be used later.

mdsage1 commented on steps/dicovdfy.cwl:
```diff
@@ -0,0 +1,40 @@
+    outputBinding:
+      glob: results.json
+
+  - id: status
```
I've commented out unused outputs that may be used later.

mdsage1 commented:
@vpchung 

Thanks for the comments and very helpful guidance. 

I've added the following changes:

1. unzip_submissions.py now creates the variables status and errors that are passed to the additional steps notify_filepath_status, add_status_annots, and check_filepath_status
2. I've commented out the outputs that aren't used in the dciovfdy.cwl, and discrepancy.cwl.

The results are generated by the scripts that are stored in the validate_score:v4 docker image. They should recieve 3 different results: scoring_results, discrepancy_results, and dciovdfy_results which are named scoring_report_series.xlsx, discrepancy_report_participant.csv, and dciodvfy_report.csv, respectively. Each are generated by 1 of 3 python scripts that calls on a group of 12 python modules written by the organizers that are all 100+ lines of code. The repo housing their scripts will automatically save the files to the working directory. I'm guessing that [this](https://github.com/Sage-Bionetworks-Challenges/brats2023/blob/main/evaluation/lesionwise/score.py#L116-L121) may not work if placed in the docker image? Also, I'd have to find out from the organizers what scripts actually create each of the final files that are generated and saved in the working directory. If the docker image is already producing the files is there an alternative way to pull them from the image and save to Synapse when they are generated?

vpchung commented:
> I'm guessing that [this](https://github.com/Sage-Bionetworks-Challenges/brats2023/blob/main/evaluation/lesionwise/score.py#L116-L121) may not work if placed in the docker image?

It should work!  In fact, that code snippet linked above is part of the Docker image used to validate and score the BraTS submissions. 

> If the docker image is already producing the files is there an alternative way to pull them from the image and save to Synapse when they are generated?

You have written the workflow so that each of those steps will return the result files as one of the step outputs, so you actually already the pieces in-place!  But as you mentioned, please check whether the Docker container's output files are available after the container stops running.

As for uploading, I think the easiest way would be to create a new workflow step that will accept those result files as its input.  For example, let's say we want to upload `dciovdfy_results` from the `create_dciovdfy_report` step:

```yaml
create_dciovdfy_report:
    run: steps/dciodvfy.cwl
    in:
      - id: config_json
        source: "#unzip_generate_config/config_json"
    out:
      - id: dciovdfy_results
      - id: results
      - id: status
```

then the new workflow step could look something like this:

```yaml
upload_to_synapse:
    run: steps/synapse_upload.cwl
    in:
      - id: synapse_config   # this input is needed so that uploading to Synapse is possible
        source: "#synapseConfig"
      - id: parent_id  # this input is needed so that Synapse knows where to upload file
        source: "#adminUploadSynId"
      - id: dciovdfy_results
        source: "#create_dciovdfy_report/dciovdfy_results"
    out:
      - id: dciovdfy_synid
```

where `synapse_upload.cwl` could look something like this:

```yaml
#!/usr/bin/env cwl-runner
cwlVersion: v1.0
class: CommandLineTool
label: Get result files then upload to Synapse

requirements:
- class: InlineJavascriptRequirement
- class: InitialWorkDirRequirement
  listing:
  - entryname: upload_results_to_synapse.py
    entry: |
      #!/usr/bin/env python
      import synapseclient
      import argparse
      import json
      import os
      import tarfile

      parser = argparse.ArgumentParser()
      parser.add_argument("--dciovdfy_file", required=True)
      parser.add_argument("-c", "--synapse_config", required=True)
      parser.add_argument("--parent_id", required=True)
      args = parser.parse_args()

      syn = synapseclient.Synapse(configPath=args.synapse_config)
      syn.login()

      results = {}
      dciovdfy = synapseclient.File(args.dciovdfy_file, parent=args.parent_id)
      dciovdfy = syn.store(dciovdfy)
      results['dciovdfy'] = dciovdfy.id
      with open('results.json', 'w') as out:
        out.write(json.dumps(results))

inputs:
- id: dciovdfy_results
  type: File
- id: parent_id
  type: string
- id: synapse_config
  type: File

outputs:
- id: results
  type: File
  outputBinding:
    glob: results.json
- id: dciovdfy_synid
  type: string
  outputBinding:
    glob: results.json
    outputEval: $(JSON.parse(self[0].contents)['dciovdfy'])
    loadContents: true

baseCommand: python3
arguments:
- valueFrom: upload_results_to_synapse.py
- prefix: --dciovdfy_file
  valueFrom: $(inputs.dciovdfy_results)
- prefix: -c
  valueFrom: $(inputs.synapse_config.path)
- prefix: --parent
  valueFrom: $(inputs.parent_id)

hints:
  DockerRequirement:
    dockerPull: sagebionetworks/synapsepythonclient:v2.7.2
```

If the above workflow code works, you can then expand it to include `scoring_results` and `discrepancy_results` as inputs too , so that they are also uploaded to Synapse.

---

Hopefully that helps, lmk if there's anything I can expand on~

vpchung commented on workflow.cwl:
```diff
@@ -92,20 +92,112 @@ steps:
+        source: "#unzip_generate_config/config_json"
+    out:
+      - id: discrepancy_results
+      - id: results
```
```suggestion
```

Since `discrepancy.cwl` is no longer returning this, this output should be removed from the general workflow as well

vpchung commented on workflow.cwl:
```diff
@@ -92,20 +92,112 @@ steps:
+    out:
+      - id: discrepancy_results
+      - id: results
+      - id: status
```
```suggestion
```

vpchung commented on workflow.cwl:
```diff
@@ -92,20 +92,112 @@ steps:
+        source: "#unzip_generate_config/config_json"
+    out:
+      - id: dciovdfy_results
+      - id: results
```
```suggestion
```

vpchung commented on workflow.cwl:
```diff
@@ -92,20 +92,112 @@ steps:
+    out:
+      - id: dciovdfy_results
+      - id: results
+      - id: status
```
```suggestion
```

mdsage1 commented:
@vpchung I've incorporated the requested changes and added logic for saving the files to synapse.

vpchung commented on steps/synapse_upload.cwl:
```diff
@@ -0,0 +1,84 @@
+  type: string
+  outputBinding:
+    glob: results.json
+    outputEval: $(JSON.parse(self[0].contents)['dciovdfy'])
```
This output will not exist, since `results.json` is being overwritten above.  That is, `dciovdfy` will not be found in the final `results.json` file.

vpchung commented on steps/synapse_upload.cwl:
```diff
@@ -0,0 +1,84 @@
+- id: synapse_config
+  type: File
+
+outputs:
```
In the main workflow CWL, I see that it is expecting 3 outputs from this step:

```yaml
  upload_to_synapse:
    run: steps/synapse_upload.cwl
    ...
    out:
      - id: dciovdfy_synid
      - id: discrepancy_synid
      - id: scoring_synid
 ```

however this CWL script indicates only two outputs: `results` and `dciovdfy_synid`

vpchung commented on steps/synapse_upload.cwl:
```diff
@@ -0,0 +1,84 @@
+      syn = synapseclient.Synapse(configPath=args.synapse_config)
+      syn.login()
+
+      dciovdfy_results = {}
+      discrepancy_results = {}
+      scoring_results = {}
+
+      dciovdfy = synapseclient.File(args.dciovdfy_file, parent=args.parent_id)
+      dciovdfy = syn.store(dciovdfy)
+      dciovdfy_results['dciovdfy'] = dciovdfy.id
+      with open('results.json', 'w') as out:
+          out.write(json.dumps(dciovdfy_results))
+
+      discrepancy = synapseclient.File(args.discrepancy_file, parent=args.parent_id)
+      discrepancy = syn.store(discrepancy)
+      discrepancy_results['discrepancy'] = discrepancy.id
+      with open('results.json', 'w') as out:
+          out.write(json.dumps(discrepancy_results))
+
+      scoring = synapseclient.File(args.scoring_file, parent=args.parent_id)
+      scoring = syn.store(scoring)
+      scoring_results['scoring'] = scoring.id
+      with open('results.json', 'w') as out:
+          out.write(json.dumps(scoring_results))
```
How can this be re-written so that `results.json` contains more than a single key-value pair?

mdsage1 commented:
@vpchung I've updated the outputs in synapse_upload.cwl to include the `scoring_synid` and `discrepancy_synid`.  I also updated the base command arguments because I noticed I didn't add the additional arguments. The python script was rewritten so the `results.json ` contains more than a single key-value pair.

vpchung commented on workflow.cwl:
```diff
@@ -79,33 +79,129 @@ steps:
+      - id: dciovdfy_results
+        source: "#create_dciovdfy_report/dciovdfy_results"
+      - id: discrepancy_results
+        source: "#create_dciovdfy_report/discrepancy_results"
```
```suggestion
        source: "#create_discrepancy_report/discrepancy_results"
```

vpchung commented on workflow.cwl:
```diff
@@ -79,33 +79,129 @@ steps:
+      - id: dciovdfy_synid
+      - id: discrepancy_synid
+      - id: scoring_synid
+      
```
Add a CWL step to annotate the submission with these synIDs?  Something similar to the `annotate_validation_with_output` step (lines 237-247).

vpchung reviewed (changes requested):
Almost there!!  Just a couple minor fixes then I think we can start some dryruns.

mdsage1 commented:
@vpchung I committed the suggested changes and added an additional step to the workflow. My VSCode connection did hiccup so please let me know if something didn't come through.

vpchung commented on workflow.cwl:
```diff
@@ -79,33 +79,147 @@ steps:
+  #   out:
+  #     - id: filepath_gold
+  
+  annotate_validation_with_output:
```
You will need to use a different name, as step names must be unique.

mdsage1 replied:
I changed the step name to `annotate_full_evaluation_with_output`

vpchung replied:
Has this commit been pushed?  I don't see the change here.

vpchung commented on workflow.cwl:
```diff
@@ -79,33 +79,147 @@ steps:
+  annotate_validation_with_output:
+    run: https://raw.githubusercontent.com/Sage-Bionetworks/ChallengeWorkflowTemplates/v4.1/cwl/annotate_submission.cwl
+    in:
+      - id: dciovdfy_synid
+        source: "#upload_to_synapse/dciovdfy_synid"
+      - id: discrepancy_synid
+        source: "#upload_to_synapse/discrepancy_synid"
+      - id: scoring_synid
+        source: "#upload_to_synapse/scoring_synid"
```
```suggestion
```

These are not inputs in the [annotate_submission CWL script](https://raw.githubusercontent.com/Sage-Bionetworks/ChallengeWorkflowTemplates/v4.1/cwl/annotate_submission.cwl), so I would remove these.

vpchung commented on workflow.cwl:
```diff
@@ -79,33 +79,147 @@ steps:
+      - id: scoring_synid
+        source: "#upload_to_synapse/scoring_synid"
+      - id: annotation_values
+        source: "#validate/results"
```
The source should be the JSON file from the `upload_to_synapse` step rather than `validate`.

mdsage1 replied:
upload_to_synapse doesn't output results. Should I add that to the output?

vpchung replied:
Makes sense to me!

vpchung commented on steps/synapse_upload.cwl:
```diff
@@ -0,0 +1,102 @@
+  type: File
+  outputBinding:
+    glob: results.json
+outputs:
```
```suggestion
```

`outputs` already defined on line 58.

mdsage1 commented:
@vpchung I've made the suggested changes.

mdsage1 commented:
@vpchung It looks like the saved changes weren't pushed in workflow.cwl. I've pushed them.

vpchung commented on steps/synapse_upload.cwl:
```diff
@@ -0,0 +1,101 @@
+  type: File
+  outputBinding:
+    glob: results.json
+  - id: dciovdfy_synid
+    type: string
+    outputBinding:
+      glob: results.json
+      outputEval: $(JSON.parse(self[0].contents)['dciovdfy'])
+      loadContents: true
+
+  - id: discrepancy_synid
+    type: string
+    outputBinding:
+      glob: results.json
+      outputEval: $(JSON.parse(self[0].contents)['discrepancy'])
+      loadContents: true
+
+  - id: scoring_synid
+    type: string
+    outputBinding:
+      glob: results.json
+      outputEval: $(JSON.parse(self[0].contents)['scoring'])
+      loadContents: true
```
I think the indentations here are off?  I could be wrong, so it might be worth to check by running this script locally.

vpchung replied:
I think the indentations here are off? I might be wrong though, so it could be worth to check by running this script locally?

vpchung commented on workflow.cwl:
```diff
@@ -79,33 +79,142 @@ steps:
+        source: "#create_discrepancy_report/discrepancy_results"
+      - id: scoring_results
+        source: "#create_scoring_report/scoring_results"
+    out:
+      - id: dciovdfy_synid
+      - id: discrepancy_synid
+      - id: scoring_synid
```
This step is missing the `results` output, which the `annotate_full_evaluation_with_output` step relies on.

mdsage1 replied:
Added the `results` output.

vpchung reviewed (changes requested):
OK, almost there, part 2! lol

mdsage1 commented:
@vpchung I've updated the results output and corrected the indentation.

vpchung reviewed (approved):
This looks good to me.  Let's do a dry-run, to ensure we didn't miss anything.
